### DISCUSSION TITLE
`cURL error 28` spamming logs (imaginary)

### DISCUSSION THREAD

**Discussion by DenuxPlays** [nextcloud/all-in-one#2295] [Questions] [answered]
### Steps to reproduce
1. Installed imaginary + recognize
2. uploaded a bunch of photos to the nextcloud

### Expected behavior
No error showing up in the nextcloud log

### Actual behavior 
cURL spamming the nextcloud logs and poor performance.
Performance notes:
This is maybe due to the fact that imaginary and recognized analyse the images and my server is hitting its limits.

### Host OS 
Fedora 37

#### Nextcloud AIO version
4.7.0

#### Current channel 
stable

#### Other valuable info 

Error:
```
[no app in context] Error: GuzzleHttp\Exception\ConnectException: cURL error 28: Operation timed out after 30001 milliseconds with 0 bytes received (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://nextcloud-aio-imaginary:9000/pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D at <<closure>>

 0. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 158
    GuzzleHttp\Handler\CurlFactory::createRejection(["GuzzleHttp\\Ha ... l], [28,"Operation t ... "])
 1. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 110
    GuzzleHttp\Handler\CurlFactory::finishError(["GuzzleHttp\\Handler\\CurlHandler"], ["GuzzleHttp\\Ha ... l], ["GuzzleHttp\\Handler\\CurlFactory"])
 2. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Handler/CurlHandler.php line 47
    GuzzleHttp\Handler\CurlFactory::finish(["GuzzleHttp\\Handler\\CurlHandler"], ["GuzzleHttp\\Ha ... l], ["GuzzleHttp\\Handler\\CurlFactory"])
 3. /var/www/html/lib/private/Http/Client/DnsPinMiddleware.php line 113
    GuzzleHttp\Handler\CurlHandler->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 4. /var/www/html/3rdparty/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php line 64
    OC\Http\Client\DnsPinMiddleware->OC\Http\Client\{closure}("*** sensitive parameters replaced ***")
 5. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 31
    GuzzleHttp\PrepareBodyMiddleware->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 6. /var/www/html/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php line 71
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 7. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 63
    GuzzleHttp\RedirectMiddleware->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
 8. /var/www/html/3rdparty/guzzlehttp/guzzle/src/HandlerStack.php line 75
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 9. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 331
    GuzzleHttp\HandlerStack->__invoke("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
10. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 168
    GuzzleHttp\Client->transfer("*** sensitive parameters replaced ***", "*** sensitive parameters replaced ***")
11. /var/www/html/3rdparty/guzzlehttp/guzzle/src/Client.php line 187
    GuzzleHttp\Client->requestAsync("post", ["GuzzleHttp\\Psr7\\Uri"], ["/mnt/ncdata/fi ... "])
12. /var/www/html/lib/private/Http/Client/Client.php line 289
    GuzzleHttp\Client->request("post", "http://nextclou ... e", ["/mnt/ncdata/fi ... "])
13. /var/www/html/lib/private/Preview/Imaginary.php line 141
    OC\Http\Client\Client->post("http://nextclou ... e", [["[{\"operation ... ]])
14. /var/www/html/lib/private/Preview/GeneratorHelper.php line 63
    OC\Preview\Imaginary->getCroppedThumbnail(["OC\\Files\\Node\\File"], 256, 256, false)
15. /var/www/html/lib/private/Preview/Generator.php line 272
    OC\Preview\GeneratorHelper->getThumbnail(["OC\\Preview\\Imaginary"], ["OC\\Files\\Node\\File"], 256, 256, false)
16. /var/www/html/lib/private/Preview/Generator.php line 151
    OC\Preview\Generator->getSmallImagePreview(["OC\\Files\\SimpleFS\\SimpleFolder"], ["OC\\Files\\Node\\File"], "image/jpeg", "", false)
17. /var/www/html/lib/private/Preview/Generator.php line 114
    OC\Preview\Generator->generatePreviews(["OC\\Files\\Node\\File"], [[64,64,false,"fill"]], "image/jpeg")
18. /var/www/html/lib/private/PreviewManager.php line 185
    OC\Preview\Generator->getPreview(["OC\\Files\\Node\\File"], 64, 64, false, "fill", null)
19. /var/www/html/apps/photos/lib/Controller/PreviewController.php line 162
    OC\PreviewManager->getPreview(["OC\\Files\\Node\\File"], 64, 64)
20. /var/www/html/apps/photos/lib/Controller/PreviewController.php line 128
    OCA\Photos\Controller\PreviewController->fetchPreview(["OC\\Files\\Node\\File"], 64, 64)
21. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 225
    OCA\Photos\Controller\PreviewController->index(28927, 64, 64)
22. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 133
    OC\AppFramework\Http\Dispatcher->executeController(["OCA\\Photos\\C ... "], "index")
23. /var/www/html/lib/private/AppFramework/App.php line 172
    OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Photos\\C ... "], "index")
24. /var/www/html/lib/private/Route/Router.php line 298
    OC\AppFramework\App::main("OCA\\Photos\\Co ... r", "index", ["OC\\AppFramewo ... "], ["28927","photos.preview.index"])
25. /var/www/html/lib/base.php line 1047
    OC\Route\Router->match("/apps/photos/api/v1/preview/28927")
26. /var/www/html/index.php line 36
    OC::handleRequest()

GET /apps/photos/api/v1/preview/28927?etag=6bb99956436b91a0b58e4c0463161f0d&x=64&y=64
from 91.1.53.169 by Denux at 2023-04-06T20:32:43+00:00
```
Screenshot from my nextcloud logs:
![image](https://user-images.githubusercontent.com/83671398/230489047-8d67b245-c7ad-42bc-9589-af07067bc4de.png)

The error message shows up about 2 times per minutes.
Also recognize is running in WASM mode if this maybe causes some problems...

I think that this is caused by nextcloud aio but I am not 100% sure.
If the issue should be reported on the imaginary repository please let me know.

**Comment by szaimen** [comment]
Hi, can you upgrade to AIO v4.8.1 and afterwards follow https://github.com/nextcloud/all-in-one/discussions/2208 to see if it fixes the problem?

**Reply by DenuxPlays** [reply]
Hey,
currently doing the update to NC26 as you said.
One question though: FTS is incompatible with NC26 or?

**Reply by szaimen** [reply]
FTS is indeed currently incompatible with nc26

**Comment by DenuxPlays** [comment]
![image](https://user-images.githubusercontent.com/83671398/232129512-209b0121-7c25-42ac-a3a7-190eb466dee3.png)
Sadly this doesn't resolved anything.
Still the same
Also a new error appeared
![image](https://user-images.githubusercontent.com/83671398/232130028-2145edf5-24a3-4ebd-98a9-7eb3e8de7b42.png)

**Reply by szaimen** [reply]
Have you been able to upgrade to Nc26 by following the above guide btw? Or are you still on nc25?

**Reply by DenuxPlays** [reply]
> Have you been able to upgrade to Nc26 by following the above guide btw? Or are you still on nc25?

I was able to upgrade. Currently running Nextcloud 26.0.1 RC 1

**Reply by szaimen** [reply]
Ah I see and with nc26 you could still reproduce the problem?

**Reply by DenuxPlays** [reply]
> Ah I see and with nc26 you could still reproduce the problem?

Yes.
The logs I showed was after upgrading to NC26.
But it only happens when some kind of "load" is there.
So uploading a lot of images, running MediaDC or recognize for example.

**Reply by DenuxPlays** [reply]
But recognize has a lot of it own problems so I wouldn't really include it.

**Comment by szaimen** [comment]
May I ask how much cpus and RAM does your server have?

**Reply by DenuxPlays** [reply]
4 cores and 8 GB of RAM
I use 2GB of RAM (including nextcloud)so that should not be a problem.

**Reply by szaimen** [reply]
Yes, that should be sufficient...

**Comment by szaimen** [comment]
> So uploading a lot of images, running MediaDC or recognize for example.

So the thing is that with nc26 we limited the amount of previews that can get generated in parallel. This should in theory solve the problem. The question is now if the limit is still too high for you cpu or if apps like recognize or mediadc are not taking the limit into account. In any way is this not AIO related and needs to be fixed in either of these apps or in Nexcloud server.

**Reply by DenuxPlays** [reply]
Ah okay thank you.

**Reply by DenuxPlays** [reply]
The CPU is not really the problem.
When imaginary generates previews the CPU never goes over 70% load. 
I try to upload a lot of images and then I'll give you an update

**Reply by szaimen** [reply]
The imaginary logs would also be pretty useful when this happens!

**Reply by DenuxPlays** [reply]
So when uploading 35 photos (236MiB) the error appeared.
CPU peaked at 85% this time.
Error appeared 9 times inside the nextcloud logs.

Logs from the imaginary container (`sudo docker logs nextcloud-aio-imaginary`):
```
172.18.0.8 - - [15/Apr/2023 11:34:57] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 30.0040
172.18.0.8 - - [15/Apr/2023 11:35:00] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9996
172.18.0.8 - - [15/Apr/2023 11:35:02] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 30.0005
172.18.0.8 - - [15/Apr/2023 11:35:30] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9870
172.18.0.8 - - [15/Apr/2023 11:35:31] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9890
172.18.0.8 - - [15/Apr/2023 11:35:31] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9985
172.18.0.8 - - [15/Apr/2023 11:35:33] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9988
172.18.0.8 - - [15/Apr/2023 11:35:35] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 30.0005
172.18.0.8 - - [15/Apr/2023 11:35:35] "POST /pipeline?operations=%5B%7B%22operation%22%3A%22autorotate%22%7D%2C%7B%22operation%22%3A%22fit%22%2C%22params%22%3A%7B%22width%22%3A256%2C%22height%22%3A256%2C%22stripmeta%22%3A%22true%22%2C%22type%22%3A%22jpeg%22%2C%22norotation%22%3A%22true%22%2C%22quality%22%3A%2260%22%7D%7D%5D HTTP/1.1" 400 41 29.9986
```

**Reply by szaimen** [reply]
Could you send me an example file that I could use for testing in order to check if I can reproduce the issue with it?

**Reply by DenuxPlays** [reply]
Yeah use my GH logo I guess
![GitHub Logo](https://user-images.githubusercontent.com/83671398/232215922-173ec473-4f04-4ae7-a3e3-b6a52dc37217.jpg)

But I dont know how imaginary handles duplicates. Maybe you have to use different photos.

**Reply by DenuxPlays** [reply]
Also previews generate very slow and my whole nextcloud is very very slow since uploading these images.
No CPU is being used:
![image](https://user-images.githubusercontent.com/83671398/232216362-a589b20f-b6ae-489c-b455-eba9a488532b.png)

And my RAM climbed from 2 GiB to 2.75 GiB but I see no reason for the slow behavior.

**Reply by szaimen** [reply]
I fear I need one of the 35 images thar you just uploaded that triggered this behaviour.

**Reply by DenuxPlays** [reply]
> I fear I need one of the 35 images thar you just uploaded that triggered this behaviour.

Yeah wait a minute.
Just let me check if their have any metadata in them that I have to remove.

**Reply by szaimen** [reply]
> Just let me check if their have any metadata in them that I have to remove

I feae that is exactly the point. I need it with metadata as it might causw the slow down..

**Reply by szaimen** [reply]
You can also send it to my mail-address.

**Reply by szaimen** [reply]
You can also send it to my mail-address.

**Reply by DenuxPlays** [reply]
I send you the image.
And I took my 15 minutes to open it....
Also my RAM usage increased to 4.35GiB so I assume that their is some kind of memory leak.

**Reply by szaimen** [reply]
Thanks, i got them! I highly suspect the autorotate feature to be the culprit here. That is why I needed the images with metadate which possibly causes the slowdown. I'll let you know if I find something.

**Reply by DenuxPlays** [reply]
Thank you :)
If I should test something please let me know

**Reply by szaimen** [reply]
I fear I cant reproduce the issue with the picture that you've send. There also doesn't seem to be any metadata in the picture so I doubt that the issue will be reproducibe with the picture on your end either.

**Reply by DenuxPlays** [reply]
Hey,
I just shared the folder with you via email.
I hope the performance is okay because I didnt restart my instance and the error causes performance issues.

**Reply by szaimen** [reply]
Thanks, I'll check with those.

**Reply by szaimen** [reply]
Cannot reproduce even with these. However I am already running the overhauled and updated imaginary container which will be included in the next release. Maybe it improves things also on your side :)

**Reply by DenuxPlays** [reply]
> Cannot reproduce even with these. However I am already running the overhauled and updated imaginary container which will be included in the next release. Maybe it improves things also on your side :)

Okay I hope.
When this didnt resolve my issue then I will completely reset my whole server and see if this will fix it.
So I keep imaginary turned off for now.

**Comment by szaimen** [comment]
The Imaginary update is now released with v4.9.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by DenuxPlays** [reply]
No sadly the same.
I guess its a problem on my side then :/

**Reply by szaimen** [reply]
The only thing that we could do from software side then would be increasing the timeout but not sure if this is a good idea...

**Reply by szaimen** [reply]
Does this also happen if you disable recognoze and mediadc?

**Reply by DenuxPlays** [reply]
> The only thing that we could do from software side then would be increasing the timeout but not sure if this is a good idea...

No I think that is not the solution because after the timeout imaginary still sent 0 bytes.

**Reply by DenuxPlays** [reply]
> Does this also happen if you disable recognoze and mediadc?

Yes

**Reply by DenuxPlays** [reply]
I also removed recognize because it isn't working for me

**Reply by szaimen** [reply]
> Yes

Hm... So yes increasing the timeout seems like the only feasible way then...

**Reply by szaimen** [reply]
However possinly you would even break that...

**Reply by szaimen** [reply]
Can you check if applying this patch resolves the issue for you?

https://patch-diff.githubusercontent.com/raw/nextcloud/server/pull/37148.patch

**Reply by DenuxPlays** [reply]
As I said maybe a server reset and new nc installation will fix it...

**Reply by DenuxPlays** [reply]
> Can you check if applying this patch resolves the issue for you?
> 
> https://patch-diff.githubusercontent.com/raw/nextcloud/server/pull/37148.patch

No still the same
![image](https://user-images.githubusercontent.com/83671398/232249730-405b506f-f01a-46f9-9843-b1fe4fc4f9a9.png)

still 0 bytes

**Reply by szaimen** [reply]
Hm are you sure the patch was correctly applied? It should report a different error then, that it timed out after 120000 ms

**Reply by DenuxPlays** [reply]
Applied the patch following the documentation:
https://docs.nextcloud.com/server/latest/admin_manual/issues/applying_patch.html

**Reply by DenuxPlays** [reply]
And it is a different error it timed out after 30.000 instead of 30.001 ms

**Reply by szaimen** [reply]
> 30.000 instead of 30.001 ms

Looks like the same error to me. If the patch was applied it should report that it timed out after 120000 or the patch is not working

**Reply by DenuxPlays** [reply]
> > 30.000 instead of 30.001 ms
> 
> Looks like the same error to me. If the patch was applied it should report that it timed out after 120000 or the patch is not working

Yeah...

**Reply by szaimen** [reply]
maybe try to restart the containers once?

**Reply by DenuxPlays** [reply]
Already tried that after I saw that it still said 30.000

**Reply by szaimen** [reply]
Then I am out of ideas unfortunately...

**Reply by DenuxPlays** [reply]
Yeah me too
Later next week I'll upgrade to Fedora 38 and then I'll completely reinstall nextcloud and try if it works out of the box.

**Reply by DenuxPlays** [reply]
> The Imaginary update is now released with v4.9.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

So as I said.
I reinstalled nextcloud. Fresh install, nextcloud aio beta (v4.9.0) and I still got the error. I did nothing.
I also switched the server from my cheap VPS to a dedicated sever with 32GB of RAM and 8 cores.
The first thing I did was tried to auto upload my pictures from my phone and then the error appeared after round about 500 pictures where uploaded.

**Comment by szaimen** [comment, marked as answer]
Please open a bug report in server then as it needs to get addressed there: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

**Reply by DenuxPlays** [reply]
Already did a while ago: https://github.com/nextcloud/server/issues/37654

**Reply by szaimen** [reply]
Ah indeed. Thanks!
